src/parseTools.js

```diff
@@ -1292,6 +1319,29 @@ function sendI64Argument(low, high) {
   return low + ', ' + high;
 }
 
+// Any function called from wasm64 may have bigint args, this function takes
+// a list of variable names to convert to number.
+function from64(x) {
+  if (!MEMORY64) {
+    return '';
+  }
+  if (Array.isArray(x)) {
+    let ret = '';
+    for (e of x) ret += from64(e);
+    return ret;
+  } else {
+    return `${x} = Number(${x});`;
```

Comment:
Why/when do `BigInts` need to be converted to numbers? Won't that splice off high bits of address values?

Reply:
Yes, in this first iteration, we will only allow 52-bits pointers. I'm not sure when we will have petabyte Wasm memories, but when that time comes, this code will need to be overhauled.

Doing that right now is impractical, since BigInts are incompatible with Numbers, so the amount of code that would need to change is substantial. Not only that, but likely we would want to keep using Number for wasm32, which means it would somehow need to be conditional. That is way too much complication for a first iteration.

Reply:
I mean, I don't expect users to have petabytes of memory, but they might do address space compartmentalization themselves and require petabytes of distinct address space, just like native applications do?

> BigInts are incompatible with Numbers

What are the actual incompatibilities you have run into so far that have necessitated the BigInt->Number conversion?

I am not arguing that it should not be done, but to learn so I can look out for such patterns when writing e.g. my Wasm WebGPU bindings library.

Reply:
These are pointers passed to JS meant to be valid pointers to Wasm memories, so any address space trickery these apps do themselves should already have been resolved before passing it to these functions.

Either way this is a first iteration of 64-bit support, not supporting invalid pointers to be passed to JS seems an entirely pragmatic limitation to me. This PR as is doesn't even implement fully working 64-bit support. Many improvements to be done in future PRs.

> What are the actual incompatibilities you have run into so far that have necessitated the BigInt->Number conversion?

100% of the current JS code represent any sort of numbers as Number. BigInt <-> Number require explicit conversion each way. Summary: a LOT of code would need modifying to support this. Even if representing these values as BigInt thruout would be a good idea, it would not be a good idea for this PR.

And like I said, there's the issue of wasm32 likely not wanting to pay the cost of BigInts everywhere, so all BigInt upgrades in codegen needing to be conditional.